### PR TITLE
feat(ClustersList): new tests for url params

### PR DESCRIPTION
**Pull request by Fewwy**
Added new tests for the ClusterList table

I had to add 2 checks to separate how urlParamConvert and hasChip work.
I think that this could be improved

### CODE REVIEW

File: cypress/utils/filters.js [hunk 1]
```diff
@@ -75,21 +78,37 @@ function applyFilters(filters, filtersConf) {
...
+  } else if (type === 'cluster') {
+    const filterCategory = _.find(
+      _.values(CLUSTER_FILTER_CATEGORIES),
+      (it) => it.urlParam === 'hits'
+    );
+    const title = _.capitalize(filterCategory.title);
+    const label = _.find(filterCategory.values, (it) => it.value === value)
+      ?.label.props.children;
+    console.log(title);
+    console.log(label);
```
Comment (adonispuente) [inline cypress/utils/filters.js:100]: these console.logs made it through

**Review by adonispuente — COMMENTED** [review]
When trying to run this, I keep getting an error about a tsconfig file not being found. When running im definitely up to date with master. I may be not setting something up properly, i definitely do not have this config file

File: cypress/utils/filters.js [hunk 2]
```diff
@@ -75,21 +78,37 @@ function applyFilters(filters, filtersConf) {
...
+    return [title, label];
+  }
 }
 
-function hasChip(name, value) {
-  cy.contains(CHIP_GROUP, name).parent().contains(CHIP, value);
+function hasChip(name, value, type) {
+  if (type === 'recommendation') {
```
Comment (ikerreyes) [inline cypress/utils/filters.js:106]: what is the difference at each branch of the if?

File: src/Components/ClustersListTable/ClustersListTable.cy.js [hunk 3]
```diff
@@ -199,6 +200,61 @@ describe('data', () => {
...
+    it('recognizes all parameters', () => {
+      const urlSearchParameters = new URLSearchParams(urlParams);
+      for (const [key, value] of urlSearchParameters) {
+        if (key == 'text') {
+          hasChip('Name', value, 'cluster');
+          cy.get('.pf-m-fill > .pf-c-form-control').should('have.value', value);
+        } else {
+          value.split(',').forEach((it) => {
+            const [group, item] = urlParamConvert(key, it, 'cluster');
+            item ? hasChip(group, item, 'cluster') : undefined;
```
Comment (ikerreyes) [inline src/Components/ClustersListTable/ClustersListTable.cy.js:245]: why is this conditional? Every parameter should be reflected in the UI, otherwise there is an error

File: cypress/utils/filters.js [hunk 4]
```diff
@@ -75,21 +78,37 @@ function applyFilters(filters, filtersConf) {
...
-  const filterCategory = _.find(
-    _.values(FILTER_CATEGORIES),
-    (it) => it.urlParam === key
-  );
-  const title = _.capitalize(filterCategory.title);
-  const label = _.find(filterCategory.values, (it) => it.value === value).label
-    .props.children;
-  return [title, label];
+function urlParamConvert(key, value, type) {
+  if (type === 'recommendation') {
```
Comment (ikerreyes) [inline cypress/utils/filters.js:82]: I do not understand why you need this distinction with an if. What are the differences in both branches? In addition, doing something specific for just 1 tests should be placed inside the test module, not in the generic

**Review by ikerreyes — COMMENTED** [review]
I do not see the point of the `if` clauses and those changes make the functions no longer generic

File: src/Components/ClustersListTable/ClustersListTable.cy.js (hunk 3 above)
Comment (gkarat) [reply src/Components/ClustersListTable/ClustersListTable.cy.js:245]: I also don't think the condition is necessary here. the check should pass for all filter/pagination values

File: cypress/utils/filters.js [hunk 5]
```diff
@@ -75,21 +78,37 @@ function applyFilters(filters, filtersConf) {
...
+    const filterCategory = _.find(
+      _.values(FILTER_CATEGORIES),
```
Comment (gkarat) [inline cypress/utils/filters.js:84]: \+ to the Iker's comment

I think, instead of adding an extra prop to this function, you could better focus on `FILTER_CATEGORIES` and `CLUSTER_FILTER_CATEGORIES` (this is the only thing these two branches are different with, right?). Can we get back to the initial version of the function but only replace one line in the following way:

```suggestion
      _.values(FILTER_CATEGORIES.concat(CLUSTER_FILTER_CATEGORIES)),
```

File: cypress/utils/filters.js (hunk 1 above)
Comment (gkarat) [reply cypress/utils/filters.js:100]: burn them!

File: cypress/utils/filters.js (hunk 2 above)
Comment (gkarat) [reply cypress/utils/filters.js:106]: \+, the two branches are identical

**Review by gkarat — CHANGES_REQUESTED** [review]

**Comment by Fewwy** [comment]
> When trying to run this, I keep getting an error about a tsconfig file not being found. When running im definitely up to date with master. I may be not setting something up properly, i definitely do not have this config file

That's very strange we don't have the typescript install in OCP Advisor

File: cypress/utils/filters.js (hunk 5 above)
Comment (ikerreyes) [reply cypress/utils/filters.js:84]: Maybe we can create an even more generic function, so that we pass the `FILTER_CATEGORIES` (in whichever form is required, i.e. `FILTER_CATEGORIES` or `CLUSTER_FILTER_CATEGORIES`). So the function will be something like:

```
function urlParamConvert(key, value, map) {
  const filterCategory = _.find(
    _.values(map),
    (it) => it.urlParam === key
  );
  const title = _.capitalize(filterCategory.title);
  const label = _.find(filterCategory.values, (it) => it.value === value).label
    .props.children;
  return [title, label];
}
```

and then we can create "partials" (sorry for the Python naming) for the different tests

```
urlParamConverter = (key, value) => urlParamConvert(key, value, FILTER_CATEGORIES)
```

Comment (gkarat) [reply cypress/utils/filters.js:84]: @ikerreyes, this is absolutely legit, from my point of view. Mine option is "the least effort," whereas yours is somewhat "generic/proper solution striving."

File: src/Components/ClustersListTable/ClustersListTable.cy.js [hunk 6]
```diff
@@ -199,6 +201,65 @@ describe('data', () => {
   });
 });
 
+const urlParamsList = [
```
Comment (ikerreyes) [inline src/Components/ClustersListTable/ClustersListTable.cy.js:204]: The test data lacks some cases in which hits are a list of values. E.g. `hits=1,2,3`

File: src/Components/ClustersListTable/ClustersListTable.cy.js [hunk 7]
```diff
@@ -199,6 +201,65 @@ describe('data', () => {
...
+    it('recognizes all parameters', () => {
+      const urlSearchParameters = new URLSearchParams(urlParams);
+      for (const [key, value] of urlSearchParameters) {
+        if (key == 'text') {
+          hasChip('Name', value, 'cluster');
+          cy.get('.pf-m-fill > .pf-c-form-control').should('have.value', value);
+        } else {
+          value.split(',').forEach((it) => {
+            const [group, item] = urlParamConvert(
+              'hits',
```
Comment (ikerreyes) [inline src/Components/ClustersListTable/ClustersListTable.cy.js:246]: Definitely, this is not something you want to do. Basically you are setting all items in the else clause to check for `hits`. That shouldn't be the case.

File: src/Components/ClustersListTable/ClustersListTable.cy.js [hunk 8]
```diff
@@ -199,6 +201,65 @@ describe('data', () => {
...
+              it,
+              CLUSTER_FILTER_CATEGORIES
+            );
+            item ? hasChip(group, item) : undefined;
```
Comment (ikerreyes) [inline src/Components/ClustersListTable/ClustersListTable.cy.js:250]: This is not a good idea, we want to test that each parameter we want to test should be there. If it is not there, then the is something wrong

File: src/Components/ClustersListTable/ClustersListTable.cy.js [hunk 9]
```diff
@@ -258,7 +319,7 @@ describe('clusters list table', () => {
     });
 
     it('applies total risk "All clusters" filter', () => {
-      hasChip('Total risk', 'All clusters');
+      hasChip('Total risk', 'All clusters', 'cluster');
```
Comment (ikerreyes) [inline src/Components/ClustersListTable/ClustersListTable.cy.js:322]: Has chips should be agnostic of the page

**Review by ikerreyes — CHANGES_REQUESTED** [review]
In my opinion, you are trying to make tests pass, but we should analyze the root cause for the failures when test are not passing as we expect.

E.g.
```
item ? hasChip(group, item) : undefined;
```
this goes against the nature of the test, as we want to ensure that each item we pass in the URL is reflected in the chips, and this test can prevent us. Maybe in the future with new parameters test just pass, even if something is wrong, because our assertions are not strong enough.

You are totally right that the current `urlParamConvert` function implementation is not correct for the use cases and it needs updated. As you are doing, passing it the configuration for each page (e.g. `urlParamConvert(key, it, FILTER_CATEGORIES);`) is, at least for now, a good solution.

I know tests are still failing, and we should go to the root cause. In this case, it is because the version parameter is not in the configuration. Probably the reason is that it is a very dynamic field that is constructed on the fly. However, if probably has a very easy to interpret mapping: the same field is the chip name and the values are directly mapped. For that, I propose to update the `urlParamConvert` function to something that accepts that schema as well.
What do you think about something like

```
function urlParamConvert(key, value, filters) {
  const filterCategory = _.find(_.values(filters), (it) => it.urlParam === key);
  let title, label;
  if (filterCategory === undefined) {
    title = _.capitalize(key);
    label = value;
  } else {
    title = _.capitalize(filterCategory.title);
    label = _.find(filterCategory.values, (it) => it.value === value).label
      .props.children;
  }
  return [title, label];
}
```
? We just separate the cases when the filter knows about the param passed and when not.

**Comment by ikerreyes** [comment]
I cannot use `hits=all` in the ClustersListTable`. Cypress fails to render, is this expected?

File: src/Components/ClustersListTable/ClustersListTable.cy.js (hunk 9 above)
Comment (Fewwy) [reply src/Components/ClustersListTable/ClustersListTable.cy.js:322]: yeah, that was a leftover from the first version. Good catch

File: src/Components/ClustersListTable/ClustersListTable.cy.js (hunk 8 above)
Comment (Fewwy) [reply src/Components/ClustersListTable/ClustersListTable.cy.js:250]: Yeah, that was my way to workaround the issue.

File: src/Components/ClustersListTable/ClustersListTable.cy.js (hunk 7 above)
Comment (Fewwy) [reply src/Components/ClustersListTable/ClustersListTable.cy.js]: I added an array 
`const clusterListFilterValues = ['hits'];`
to store values and utilized _.find() lodash method to search the values in the array

File: src/Components/ClustersListTable/ClustersListTable.cy.js (hunk 9 above)
Comment (ikerreyes) [reply src/Components/ClustersListTable/ClustersListTable.cy.js]: @Fewwy you still have this code of `hasChip` with 3 parameters in many other places

Comment (Fewwy) [reply src/Components/ClustersListTable/ClustersListTable.cy.js]: Removed them everywhere

File: cypress/utils/filters.js [hunk 10]
```diff
@@ -75,21 +73,27 @@ function applyFilters(filters, filtersConf) {
   }
 }
 
-function urlParamConvert(key, value) {
+function urlParamConvert(key, value, filters) {
   const filterCategory = _.find(
-    _.values(FILTER_CATEGORIES),
-    (it) => it.urlParam === key
+    _.values(filters),
+    (it) => it?.urlParam === key
```
Comment (ikerreyes) [inline cypress/utils/filters.js:79]: I think you can remove the `?` here as well

File: src/Components/RecsListTable/Downloadstuff.js [hunk 11]
```diff
@@ -0,0 +1,15 @@
+import { useGetRecsQuery } from '../../Services/SmartProxy';
```
Comment (ikerreyes) [inline src/Components/RecsListTable/Downloadstuff.js:1]: why is this file needed? What is its purpose?

File: src/Components/ClustersListTable/ClustersListTable.cy.js (hunk 6 above)
Comment (Fewwy) [reply src/Components/ClustersListTable/ClustersListTable.cy.js:204]: I made a separate Jira card covering this issue

**Review by ikerreyes — APPROVED** [review]

**Comment by Fewwy** [comment]
> 

The issue is that the CLUSTER_FILTER_CATEGORIES all categories has a different structure. I think that should be covered in the other task

**Review by gkarat — APPROVED** [review]
vincent-price-looks-good-to-me
